### ISSUE TITLE
Cannot update the masterVariant when the variantId is not 1

### ISSUE THREAD

**Issue by VasumathiJayakumar** [sphereio/sphere-node-product-csv-sync#189]
### Description
When the masterVariant has a variantId apart from 1, the importer is throwing an error.

`[ '[row 2] We need a product before starting with a variant!' ]`

### Expected Behavior
Should update the masterVariant with the appropriate variantId
### Current Behavior
* When the masterVariant has variantId is not 1, the importer is throwing an error.
* When the same masterVariant is updated with variantId 1, the update is succeed. But actually the variantId is not 1. The module seems to assume that the masterVariant will always have an ID of 1, which is not always the case (specifically after running the [`changeMasterVariant`](https://dev.commercetools.com/http-api-projects-products.html#change-master-variant) update action)
### Context

Makes it impossible to update the master variant of some products in a project

**Comment by Oehmi** [comment]
This should be done until end of February for a customer.
FYI @junajan  @wizzy25

**Comment by wizzy25** [comment]
👍

**Comment by wizzy25** [comment]
This issue has been reported and is still valid

**Comment by Oehmi** [comment]
We've discussed with @wizzy25 and @VasumathiJayakumar three options. As we'd like to go forward with option 1, we'd appreciate your input whether it's manageable for the projects you are working on @mitsch2000 @MelNiermann @ValeSauer @butenkor . Thanks!

**Option 1**
Differentiation between products is done by evaluating the product _key_ (key is optional); 
CON: Not all current projects use the product key, therefore these projects depending on the Product Import would have to update the key after this fix. 

**Option 2**
Add a _new field (boolean)_ to the import which states whether the respective variant is the master variant or not. 
CON: breaking change

**Option 3**
Differentiation between products is done by evaluating the product _name_; whenever a new name is detected, the importer concludes that a new product is given and the first variant is the master variant. 
CON: name is not unique, therefore it could be error prone

**Comment by butenkor** [comment]
I am not sure i exactly understand the issue. Was the issue already addressed here or it was different one: https://github.com/sphereio/sphere-node-product-csv-sync/pull/192?

If i understand it right in Option 1 every variant row would need to provide mandatory product key value which will allow to find all variants of same product? How masterVariant will be determined in this case - by `variantId = 1`? How masterVariant is identified if variants are matched by SKU and not variantId (which would not be supplied)?

In general we do not use sphere-node-product-csv-sync for import. I know one case for data feed export.

Are there any plans for usage of [sphere-product-import](https://github.com/sphereio/sphere-product-import) as dependency here? sphere-node-product-csv-sync would take care of mapping from CSV/XLS to JSON only and sphere-product-import would do the matching and execute update/create?

In sphere-product-import we use SKU for matching only. SKU is unfortunately optional in API but all customers have unique variant ID/SKU as it is required for inventory (and eventually later for price endpoint) linking. Usage of product/variant keys for matching might introduce more issue than expected as both key and SKU are unique and most probably will be the same (data duplication). If they will not be the same then there might be conflicts. Example:
```Existing product -> product_key1, variant_sku1```
```New product -> product_key2, variant_sku1```

This would lead to an attempt of failed product creation with `product_key2, variant_sku1`. Usage of keys for product/variants might make sense when we have dedicated variants endpoint.

We try to address/solve most known conflicts within [commercetools-node-variant-reassignment](https://github.com/commercetools/commercetools-node-variant-reassignment) module, which soon will be integrated into sphere-product-import.

**Comment by wizzy25** [comment]
> I am not sure i exactly understand the issue. Was the issue already addressed here or it was different one: #192?

It's the same issue as was tried to fix in #192, but that didn't resolve it. [This function](https://github.com/sphereio/sphere-node-product-csv-sync/blob/master/src/coffee/validator.coffee#L207) checks that a `variantId` is 1, then assigns it as the masterVariant (and the beginning of a new product). However, with the introduction of the `changeMasterVariant` update action, the `masterVariantId` is not guaranteed to always be 1, and so the module does not know what the masterVariant is at parse time. The idea of the key was to implement something on the product level that will be unique which can serve to differentiate products, with the first row (regardless of the variantId) serving as the masterVariant

> If i understand it right in Option 1 every variant row would need to provide mandatory product key value which will allow to find all variants of same product? How masterVariant will be determined in this case - by `variantId = 1`? How masterVariant is identified if variants are matched by SKU and not variantId (which would not be supplied)?

No. The supplied key is the `productKey` and not the `variantKey`, so it's supplied on the same level as the `name` or `slug`. MasterVariant will be determined by being the first row in the products

> Are there any plans for usage of [sphere-product-import](https://github.com/sphereio/sphere-product-import) as dependency here? sphere-node-product-csv-sync would take care of mapping from CSV/XLS to JSON only and sphere-product-import would do the matching and execute update/create?

Currently there are no plans. This approach could be given more thought, especially with the [variant-reassignment](https://github.com/commercetools/commercetools-node-variant-reassignment) coming soon.

> In sphere-product-import we use SKU for matching only. SKU is unfortunately optional in API but all customers have unique variant ID/SKU as it is required for inventory (and eventually later for price endpoint) linking. Usage of product/variant keys for matching might introduce more issue than expected as both key and SKU are unique and most probably will be the same (data duplication). If they will not be the same then there might be conflicts. Example:
```Existing product -> product_key1, variant_sku1```
```New product -> product_key2, variant_sku1```
This would lead to an attempt of failed product creation with `product_key2, variant_sku1`. Usage of keys for product/variants might make sense when we have dedicated variants endpoint.

This should not be a problem (I think). The SKU is on the variant level, and the variantKey is still not required (could be left out or duplicated, doesn't matter). Again, it's the `productKey` that will be required

> We try to address/solve most known conflicts within [commercetools-node-variant-reassignment](https://github.com/commercetools/commercetools-node-variant-reassignment) module, which soon will be integrated into sphere-product-import.

The actual problem is from the way this module parses the CSV to JSON, and not the importer itself

**Comment by ValeSauer** [comment]
I would prefer option 1. From my point of view utilization of keys becomes more and more mandatory if you use syncs, imports and so on anyway - so it fits into our overall strategy.

**Comment by Oehmi** [comment]
@mitsch2000 and @MelNiermann, could you please also give us your feedback on the options in order for us to take next steps? Especially in regards to option 1, which would use the product key and would need to be updated if not yet used in your projects. 

Thanks! 

FYI @wizzy25

**Comment by MelNiermann** [comment]
Hey @Oehmi 

does this mean, that the product key will become mandatory for the CSV import? 

Because if this is the case, we can't change this so easily, since the marketplace has a few partners which are using the CSV Connector and import all data via CSV files. 

We have to inform all partners first, so they have to change the implementation on there side, this means either we need a bit of time to inform partners and five them the chance to chance the imports  exports or the import can handle both (imports with keys and without). 

Best
Melanie

**Comment by wizzy25** [comment]
@MelNiermann would it be possible to get a timeframe for informing the partners?

**Comment by Oehmi** [comment]
FYI @MelNiermann : If you are using CLI, then you won't be affected unless you update to the newest version. 
However, on Impex the changes will break. Most likely, the next Impex update with these changes will be in approx. 3 weeks.

**Comment by Oehmi** [comment]
@stmeissner @VasumathiJayakumar @MelNiermann 
We will put this issue on hold for now and not deploy any changes. If you are aware of any other customer that this causes problems for, please do let us know. 

@wizzy25 Regarding the partners, they had not been informed so far.

**Comment by wizzy25** [comment]
The "fix" for this issue still exists on [this branch](https://github.com/sphereio/sphere-node-product-csv-sync/tree/189-master-variant-not-1) if it ever becomes necessary again to be implemented

**Comment by skyriakou** [comment]
@stmeissner Is this still relevant issue? Thanks

**Comment by stmeissner** [comment]
@skyriakou This issue is now been handled by another tool: https://github.com/commercetools/commercetools-node-variant-reassignment
Hence I am closing this.
